### ISSUE TITLE
Punycoder for internationalized domain names

### ISSUE THREAD

**Issue by BreyndotEchse** [zendframework/zf2#7701]
ZF needs a punycoder to allow the conversion of UTF-8 domains in IDNA ASCII form and vice versa. PHP offers two functions (`idn_to_ascii` & `idn_to_utf8`) but require intl extension to work.
The current handling of IDN is rather messy:

- Zend\Validator\Hostname: Hat it's own decoder (`decodePunycode` single responsibility?). Validation works without intl.
- Zend\Validator\Mail: Uses intl function for conversion. Returns false negatives with ext-intl=off.
- Zend\Mail (Zend\Mail\Address): No conversion at all. Sending E-Mails to internationalized domains is impossible without custom encoding.
- Zend\Mvc\Router\Http\Hostname: Does not match the UTF-8 domain. Only works with the IDNA ASCII form.

In my opinion there must be a way to handle IDN without intl and that's why I am here to ask you guys if you agree with me and where you think I should place the Punycoder.
- zendframework/zend-i18n: i18n = internationalization = the "I" in IDN
- zendframework/zend-stdlib: IDN is just a way to encode UTF-8 strings to ASCII. With `StringUtils` there is already an encoder (wrapper, at least).
- new component: Could be but should not..
- or somewhere else?

I would go with zend-stdlib because it is already required by most components.

**Comment by GeeH** [comment]
I would say it should go into `Zend\I18n` because it then keeps our `intl` extension dependencies in one place.

 > I have no idea what I'm talking about with regards to what you are saying technically

**Comment by macnibblet** [comment]
I agree with @GeeH (Does not happen very often) and I do know what you are talking about. Even tho it might feel a bit strange to have it in the `Zend\I18n` I still think it's the best place. 

The reason I don't want to see this in `Zend\Stdlib` is because it's far to specific to be "standard library"

I could see it in it's own component if adding more components doesn't add to much overhead @weierophinney ?

**Comment by weierophinney** [comment]
I think either of:

- zend-i18n
- new component

make sense. I'd go for zend-i18n if you were to use the `idn_to_ascii`/`idn_to_utf8` functionality, a standalone component if it were to have no dependencies and be a standalone implementation. This latter appeals to me as well if it were to be used in zend-mail, zend-router, and zend-validator, as that way we could keep the i18n dependency out of them.

**Comment by adamlundrigan** [comment]
A new component to polyfill [the IDN functions](http://php.net/manual/en/ref.intl.idn.php) if they're missing.  It's not really a polyfill in the usual sense, since it's filling a missing PECL dependency and not a missing language feature.

@weierophinney if this (polyfill + updating identified components to use it) is a useful addition I can take it on.

**Comment by weierophinney** [comment]
@adamlundrigan Go for it!

**Comment by BreyndotEchse** [comment]
You might take a look at this: https://github.com/phlylabs/idna-convert
Looks pretty solid

**Comment by Ocramius** [comment]
@BreyndotEchse two massive problems there:

 * [LGPL2+](https://github.com/phlylabs/idna-convert/blob/976b394b5de6f0df418195d0b30db22896cf8289/composer.json#L4)
 * [no tests](https://github.com/phlylabs/idna-convert/tree/976b394b5de6f0df418195d0b30db22896cf8289/test)

**Comment by adamlundrigan** [comment]
@Ocramius how do you feel about this one:  https://github.com/true/php-punycode

- MIT licensed
- Has some tests

With this we don't need to write our own punycode functions, and the new component essentially just proxies calls when the IDN functions are missing:

 - idn_to_ascii → TrueBV\Punycode::decode
 - idn_to_utf8 → TrueBV\Punycode::encode

To set that up, I plan to borrow the structure from the Symfony polyfill components (eg: https://github.com/symfony/polyfill-mbstring) and set up Travis to run the tests with both `intl` present and absent to ensure consistent behaviour.

Once that's done I'll tackle all the components with use IDN in some way and open an issue on each to decide how to proceed with switching those over to using the IDN functions.
